src/node_worker.cc

```diff
@@ -88,72 +84,94 @@ Worker::Worker(Environment* env,
                 env->message_port_string(),
                 parent_port_->object()).FromJust();
 
-  array_buffer_allocator_.reset(CreateArrayBufferAllocator());
-
-  CHECK_EQ(uv_loop_init(&loop_), 0);
-  isolate_ = NewIsolate(array_buffer_allocator_.get(), &loop_);
-  CHECK_NOT_NULL(isolate_);
-
-  {
-    // Enter an environment capable of executing code in the child Isolate
-    // (and only in it).
-    Locker locker(isolate_);
-    Isolate::Scope isolate_scope(isolate_);
-    HandleScope handle_scope(isolate_);
-
-    isolate_data_.reset(CreateIsolateData(isolate_,
-                                          &loop_,
-                                          env->isolate_data()->platform(),
-                                          array_buffer_allocator_.get()));
-    if (per_isolate_opts != nullptr) {
-      isolate_data_->set_options(per_isolate_opts);
-    }
-    CHECK(isolate_data_);
-
-    Local<Context> context = NewContext(isolate_);
-    Context::Scope context_scope(context);
-
-    // TODO(addaleax): Use CreateEnvironment(), or generally another public API.
-    env_.reset(new Environment(isolate_data_.get(), context));
-    CHECK_NOT_NULL(env_);
-    env_->set_abort_on_uncaught_exception(false);
-    env_->set_worker_context(this);
-    thread_id_ = env_->thread_id();
-
-    env_->Start(env->profiler_idle_notifier_started());
-    env_->ProcessCliArgs(std::vector<std::string>{},
-                            std::vector<std::string>{});
-    // Done while on the parent thread
-    AddWorkerInspector(env, env_.get(), thread_id_, url_);
-  }
-
-  // The new isolate won't be bothered on this thread again.
-  isolate_->DiscardThreadSpecificMetadata();
-
-  wrap->Set(env->context(),
-            env->thread_id_string(),
-            Number::New(env->isolate(), static_cast<double>(thread_id_)))
+  object()->Set(env->context(),
+                env->thread_id_string(),
+                Number::New(env->isolate(), static_cast<double>(thread_id_)))
       .FromJust();
 
-  Debug(this,
-        "Set up worker at %p with id %llu",
-        static_cast<void*>(this),
-        thread_id_);
+#if NODE_USE_V8_PLATFORM && HAVE_INSPECTOR
+  inspector_parent_handle_ =
+      env->inspector_agent()->GetParentHandle(thread_id_, url);
+#endif
+
+  Debug(this, "Preparation for worker %llu finished", thread_id_);
 }
 
 bool Worker::is_stopped() const {
   Mutex::ScopedLock stopped_lock(stopped_mutex_);
   return stopped_;
 }
 
+class WorkerThreadData {
```

Comment:
It is not very obivous to me what goes into `Worker` and what goes into `WorkerThreadData`. For example,
 - I would expect worker to contain fields that it can construct from within (isolate is a field, but is populated from outside)
 - many thread and synchronization data fields sit in worker, while workerthreaddata has very few, and yet those are required in its contructorsand destructor.

So: is it possible to derive a better relation (Container -> Component or Proxy etc.) between Worker and WorkerThreadData, than mutual `friendship`?

Reply:
`WorkerThreadData` contains fields that are only relevant to the thread itself, I’ve added a comment about that.